### PR TITLE
Added 'build label propagation' e2e test

### PR DESCRIPTION
**Pull request by aorcholski**
# Description

The test should ensure that build label propagation feature works properly. It consists of five steps:
1) build labels propagation is disabled
2) default build labels are propagated only
1) all custom build labels are propagated
1) existing environment variables are preserved
1) invalid propagation (empty value of build variables due to invalid references to build labels)

Additionally

test/e2e/buildlabels target has been added

## How can this be tested?
1. Create testdata/secrets/single-tenant.yaml (see test/testdata/secrets-samples).
2. Use make command to start the test.
```
make test/e2e/buildlabels
```
## Checklist
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: test/sampleapps/sampleapps.go [hunk 1]
```diff
@@ -18,28 +18,30 @@ const (
...
+			conditions.New(resource).DeploymentConditionMatch(
+				&sampleDeployment, appsv1.DeploymentAvailable, corev1.ConditionTrue)))
 
-	return ctx
+		return ctx
+	}
 }
 
-func Get(t *testing.T, ctx context.Context, resource *resources.Resources) corev1.PodList {
+func Get(t *testing.T, ctx context.Context, resource *resources.Resources, namespaceName string) corev1.PodList {
```
Comment (gkrenn) [inline test/sampleapps/sampleapps.go:42]: Should we move that to the kubeobjects pod package? - To stay more consistent with other objects e.g. daemonsets and deployments. There we already have a `get.go`

Comment (aorcholski) [reply test/sampleapps/sampleapps.go]: sampleapps.Get renamed to pod.List

File: hack/make/tests/e2e.mk [hunk 2]
```diff
@@ -12,3 +12,7 @@ test/e2e/cloudnative:
 ## Runs Application Monitoring e2e test only
 test/e2e/applicationmonitoring:
 	go test -v -tags e2e -count=1 -failfast ./test/appmon
+
+## Runs build variables e2e test only
+test/e2e/buildlabels:
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply hack/make/tests/e2e.mk:21]: would make sense

File: test/buildlabels/install.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+func assertReferences(t *testing.T, pod *corev1.Pod, expectedVariables map[string]buildLabel) {
+	require.NotNil(t, pod)
+	require.NotNil(t, pod.Spec)
+
+	appContainer := pod.Spec.Containers[0]
+	require.Equal(t, "myapp", appContainer.Name)
+
+	variablesFound := map[string]bool{}
+
+	for _, ev := range appContainer.Env {
```
Comment (meik99) [inline test/buildlabels/install.go:190]:
```suggestion
	for _, containerEnvVar := range appContainer.Env {
```

File: test/buildlabels/install.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		if value, ok := expectedVariables[ev.Name]; ok {
```
Comment (meik99) [inline test/buildlabels/install.go:191]:
```suggestion
        label, hasLabel := expectedVariables[ev.Name]
		if hasLabel {
```

File: test/buildlabels/install.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+				require.NotNil(t, ev.ValueFrom, "%s:%s pod - %s variable has empty ValueFrom property", pod.Namespace, pod.Name, ev.Name)
+				require.NotNil(t, ev.ValueFrom.FieldRef, "%s:%s pod - %s variable has empty FieldRef property", pod.Namespace, pod.Name, ev.Name)
+				assert.Equal(t, value.reference, ev.ValueFrom.FieldRef.FieldPath, "%s:%s pod - %s variable has invalid value reference", pod.Namespace, pod.Name, ev.Name)
+				variablesFound[ev.Name] = true
+			}
+		}
+	}
+
+	for name, value := range expectedVariables {
+		_, ok := variablesFound[name]
```
Comment (meik99) [inline test/buildlabels/install.go:202]:
```suggestion
		_, hasLabel := variablesFound[name]
```

File: test/buildlabels/install.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+			assertReferences(t, &podItem, expectedVariables)
+
+			assertValues(t, environmentConfig.Client().RESTConfig(), podItem, expectedVariables)
+		}
+
+		return ctx
+	}
+}
+
+func assertReferences(t *testing.T, pod *corev1.Pod, expectedVariables map[string]buildLabel) {
```
Comment (meik99) [inline test/buildlabels/install.go:181]:
```suggestion
func assertReferences(t *testing.T, pod *corev1.Pod, expectedBuildLabels map[string]buildLabel) {
```

File: test/buildlabels/install.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+		_, ok := variablesFound[name]
+		if value.reference == "" {
+			assert.False(t, ok, "%s:%s pod - %s variable found", pod.Namespace, pod.Name, name)
+		} else {
+			assert.True(t, ok, "%s:%s pod - %s variable not found", pod.Namespace, pod.Name, name)
+		}
+	}
+}
+
+func assertValues(t *testing.T, restConfig *rest.Config, podItem corev1.Pod, expectedVariables map[string]buildLabel) {
```
Comment (meik99) [inline test/buildlabels/install.go:211]:
```suggestion
func assertValues(t *testing.T, restConfig *rest.Config, podItem corev1.Pod, expectedBuildLabels map[string]buildLabel) {
```

**Review by meik99 — CHANGES_REQUESTED** [review]

File: test/buildlabels/install.go (hunk 4 above)
Comment (aorcholski) [reply test/buildlabels/install.go:191]: renamed

File: test/buildlabels/install.go (hunk 6 above)
Comment (aorcholski) [reply test/buildlabels/install.go:181]: renamed

File: test/buildlabels/install.go (hunk 5 above)
Comment (aorcholski) [reply test/buildlabels/install.go:202]: renamed

File: test/buildlabels/install.go (hunk 7 above)
Comment (aorcholski) [reply test/buildlabels/install.go:211]: renamed

File: test/buildlabels/install.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+		return ctx
+	}
+}
+
+func assertReferences(t *testing.T, podItem *corev1.Pod, expectedVariables map[string]buildLabel) {
+	require.NotNil(t, podItem)
+	require.NotNil(t, podItem.Spec)
+
+	appContainer := podItem.Spec.Containers[0]
+	assert.Equal(t, "myapp", appContainer.Name)
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply test/buildlabels/install.go]: you are right. It should be 'require' condition

File: test/buildlabels/install.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+		dtReleaseProduct:      {"metadata.labels['my-product']", "my-product"},
+		dtReleaseStage:        {"metadata.labels['my-stage']", "my-stage"},
+		dtReleaseBuildVersion: {"metadata.labels['my-build-version']", "my-build-version"},
+	}
+
+	invalidCustomBuildLabels = map[string]buildLabel{
+		dtReleaseVersion: {"metadata.labels['app.kubernetes.io/version']", "app-kubernetes-io-version"},
+		dtReleaseProduct: {"metadata.labels['app.kubernetes.io/part-of']", "app-kubernetes-io-part-of"},
+		// invalid name of STAGE label -> reference exists but actual label doesn't exist otherwise value would be "my-domain-stage"
+		dtReleaseStage: {"metadata.labels['my.domain/stagee']", ""},
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply test/buildlabels/install.go]: true

File: hack/make/tests/e2e.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (aorcholski) [reply hack/make/tests/e2e.mk]: I will merge both tests.

**Review by 0sewa0 — APPROVED** [review]

**Review by meik99 — APPROVED** [review]

**Pull request merged**